### ISSUE TITLE
millerRabinV performance enhancement?

### ISSUE THREAD

**Issue by grandpascorpion** [cartazio/arithmoi#22]
millerRabinV : This routine in Math/NumberTheory/Primes/Testing/Probabilistic.hs
seems quite inefficient.  According to the comments in the code, verifying a prime is actually "prime" with this routine would require testing 196 bases (primes through 1200).

In the last few years, there has been a project: https://miller-rabin.appspot.com/
to identify the maximum accurate solution for # of bases: 1 - 7.
A set of 7 bases exists (2, 325, 9375, 28178, 450775, 9780504, 1795265022) 
that can verify the primality of a number  <= 2^64.

Secondly, I wonder if this logic could be used to improve the performance of isPrime for numbers < 2^64.  Not sure what the cutoff would be.  http://www.trnicely.net/misc/bpsw.html

[Edit: Fixed a few typos]

**Comment by Bodigrim** [comment]
Sounds reasonable, thank you for links. Sorry, I am busy at the moment. It will be great if you have spare time to prepare pull request.

**Comment by grandpascorpion** [comment]
Sorry for the noobie question but which branch should I pull from (710+ testing) ?

**Comment by Bodigrim** [comment]
Pull `master` branch. Other branches are outdated and should be pruned in future.

**Comment by grandpascorpion** [comment]
Thanks!

**Comment by grandpascorpion** [comment]
Shoot, I missed something obvious, that one can specify the number of bases to millerRabinV.  I think a new function would make more sense than updating this one.  Otherwise, the "optimal" benefits would result in a clunky implementation. 

I think that logic could still be adapted for isPrime if there are performance benefits.

**Comment by Bodigrim** [comment]
A new function is not a problem, if it is better than old one. Maybe it is worth to run Miller-Rabin over 7 bases above first and then over specified in the `k` parameter number of consequent primes.

I encourage you to change `isPrime` and make benchmarks. It sounds like a very interesting experiment.

**Comment by grandpascorpion** [comment]
Hmm, I think the new function could just accept n as the sole parameter (if
the intention is to check the minimal set of bases).

Maybe we could do something else, like a hybrid function that combined
"new" and millerRabin,
we could ignore the k (# of bases) param if n <= 2^64 or k <8 and n >
2^64.  Otherwise, we could  "supplement" the 7 base list with other small
primes not in the list.

For instance, the optimal 7 base set is:

2,  325,  9375,  28178, 450775, 9780504, 1795265022
Say we tested the function with n = 2^100 + 3 and k = 9:
It would test with: 2, 325,  9375,  28178, 450775, 9780504, 1795265022, 3, 5

Or maybe that's just too obscure.

Anyways, I'll try to get to the benchmarking soon.

Best, Fred

Sent from my iPhone

On Mar 16, 2016, at 4:33 PM, Andrew Lelechenko <notifications@github.com>
wrote:

A new function is not a problem, if it is better than old one. Maybe it is
worth to run Miller-Rabin over 7 bases above first and then over specified
in the k parameter number of consequent primes.

I encourage you to change isPrime and make benchmarks. It sounds like a
very interesting experiment.

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/cartazio/arithmoi/issues/22#issuecomment-197536297>

**Comment by grandpascorpion** [comment]
Since these "optimal sets of bases" do have bases that may be greater than input n, this idea is not guaranteed to work properly (for most of the optimal base sets) with the current implementation of isStrongFermatPP.
(Anecdotally, it seems that if b is a multiple of prime n, the function incorrectly returns false.  I don't know if there are other scenarios that return the wrong answer when b > n and n is valid input)

**Comment by grandpascorpion** [comment]
I reached out to Wozciech Izykowski who maintains the Miller-Rabin record site and he pointed me to his implementation(s):  https://github.com/wizykowski/miller-rabin

I think I have a good solution for the solution for our "dilemma":

Two key points in his code:
if b > n, let b = b % n
if b = 0, this is a non-test. Continue to the next base in the list.  If there are no more in the list, return true(!).

I think this would work in the context of tiered checking that https://miller-rabin.appspot.com/ is promoting.  The only false positive would occur where the gcd of the bases (or a single base) is a number which has a composite factor (which is a multiple of two or more primes, not including 2) in the checking range, which would mean it would be "skipped" for each base and then return true by default.

For instance, the best known single base: 9345883071009581737 factors to: 47×98207×2024790248953.
It has been vetted there through n < 341531.  This base has no composite factors < 341531 so we are ok on that front.  The two prime factors in range: 47 and 98207 will be skipped because b%n = 0  but work properly because we are defaulting to true.

The optimal two base set is: 336781006125, 9639812373923155 which gcd of 5, which is prime of course.
(I check the optimal 3-7 base sets as well and each set's gcd is 1)

Long story short, I think a new standalone function (isPrimeTieredCheck, maybe) based on that website's findings would be a nice addition.  (It would not alter any existing code).  I'll get working on that.

**Comment by Bodigrim** [comment]
Sounds good. Feel free to ask any questions, especially regarding tests and benchmarks.

**Comment by grandpascorpion** [comment]
Thanks, I appreciate it

**Comment by Bodigrim** [comment]
@grandpascorpion did you encounter any difficulties, writing this new `tieredPrimeTest`?

**Comment by grandpascorpion** [comment]
Hi Andrew,

I found it to be a mixed bag.  Some bit ranges were faster, some slower
(than isPrime).  I can send my findings if you like

Not sure how useful it is.

Thanks,
Fred

**Comment by Bodigrim** [comment]
Yes, please. It would be even better, if you fork `arithmoi` repo and push your changes on GitHub.

**Comment by cartazio** [comment]
Indeed.  Pretend this is some sort of nasa research engineering thing. And
or that some arithmoi users actually use the lib to prototype crypto algs.
So let's try to perpetually raise the Bar of quality and aweosme.

Performance and design experiments are a great use of your own forks and
personal branches.  A progamatic experiment journal others can replicate or
compare or run!

**Comment by Bodigrim** [comment]
@grandpascorpion feel free to contact me, it you have any technical difficulties with git/github stuff.

**Comment by Bodigrim** [comment]
I decided to use GMP's implementation of Miller-Rabin test. It chooses bases at random, but it is very fast. I have also added a link to the best choice of bases for strong Fermat test.
